### ISSUE TITLE
Using 6th notebook for Video captioning

### ISSUE THREAD

**Issue by kushalj001** [bentrevett/pytorch-seq2seq#85]
I am trying to replicate the 6th notebook for video captioning. The main change required to do so would be;
__The source here would be video-frames instead of text/words__. I have truncated or padded the input videos to ensure that all the videos have 32 frames (i.e. 32 images) only. That is the `src_len` is 32. Hence the `src` shape would be `[batch_size,  src_len, 3, 224, 224]` instead of `[batch_size, src_len]` only.
However, I am not able to figure out how can I create the `src_mask` for the 5-dimensional tensor, since the energy vector calculated in the `MultiHeadAttentionLayer` is a 4-dimensional tensor. I have padded the videos whose number-of-frames are less than 32 with zeroes.
For target, the framework remains exactly the same.
Any ideas? 
Also is this idea viable?
Thank you.

**Comment by bentrevett** [comment]
I'll admit I don't know how to apply the transformer for video captioning, but it looks like the idea is viable as there has been work on it in the past, see: https://arxiv.org/pdf/1804.00819.pdf

For masking, what you want to do is make your mask broadcastable. For example if you are trying to mask a sequence of `[batch_size, 32, 3, 224, 224]` then your mask should be `[batch_size, 32, 1, 1, 1]` 

Let's say you had a batch size of 2, and your first video is actually 32 frames long and the second is only 20, then your mask should have the first row be all 1s and the second row be twenty 1s and then twelve 0s. You can then just unsqueeze this 3 times to add the extra dimensions so it is broadcastable. 

If you want to split your tensor into `n_heads`, then it would be: `[batch_size, n_heads, src_len, n_channels, height, width]` and a valid mask for that would be `[batch_size, 1, src_len, 1, 1, 1]`. As long as your mask is correct, you can just unsqueeze it and it will broadcast itself to mask out the padded frames. 

How are you calculating the `energy` between the query and the key for your videos?

**Comment by kushalj001** [comment]
Thank you for replying. 
I'll explain what my hypothesis is. Instead of the input source text, my input would be video frames. 
As the source words are converted into a continuous representation by passing them through an embedding layer, similarly, I am converting the frames by passing them through a pretrained google lenet network. To be more precise, I am using the weights of layer pool 5 to encode the frames into a continuous representation. This gives me the following transformation:
`[batch_size, frames (32), 3, 224, 224]` to `[batch_size, 32, 1024]`. I further downsample this by passing it through a linear layer such that the output is `[batch_size, 32, 512]` which represents the output of the embedding layer in case of machine translation. 
The calculation of energy is exactly where I was stuck due to the difference in tensor sizes. `energy` is a 4-dimensional tensor, whereas my `src` would be a 5-dimensional vector. 
Wouldn't calculating energies in the same work? I am basically trying to enforce/hypothesize that changing the source encoding would work for video captioning as well as it does for machine-translation.

**Comment by bentrevett** [comment]
If your down-sampling is `[batch size, 32, 512]` then can you just not use this as the input to the multi-headed attention layer? This would mean you are performing self-attention on the down-sampled images but that might be sufficient?

**Comment by kushalj001** [comment]
Hi @bentrevett 
Sorry for coming back so late.


> If your down-sampling is `[batch size, 32, 512]` then can you just not use this as the input to the multi-headed attention layer? This would mean you are performing self-attention on the down-sampled images but that might be sufficient?

Even I believed this would work. But I am not able to figure out the source mask's shape. As you said above it should be broadcastable and for a sequence of shape `[batch_size, 32, 3, 224, 224]`, my src_mask should be `[batch_size, 32, 1, 1, 1]`. In order to transform my `src` tensor into this mask, should I drop the trailing dimensions and `unsqueeze`  the tensor with 3 `1s`? 

Even then, the dimension of my energy would be `[bs, num_heads, query_len, key_len]` which is a 4-d tensor and my mask would be 5-d tensor. I am not able to figure out how things would work here.
Thank you.

**Comment by yugaljain1999** [comment]
you can just expel out last dimension of src_mask as it is 1 and then you'll get 4 dimensional `'src_mask` tensor.
